### ISSUE TITLE
Have access to target injection point from Providers

### ISSUE THREAD

**Issue by mmanciop** [google/guice#948]
I find myself often in the situation that what I would like to inject should depend on the values inside BindingAnnotations on the members to be injected. Guice can deal with that when all possible values are known upfront and bound in advance against implementations of the BindingAnnotations that implement equals and hashcode as specified by the JDK APIs.

However, the method of implementing BindingAnnotations and bind them in advance does not work when the possible values are not known a priori. Assisted injection is a (IMO, less elegant) alternative in many such cases, except when the type to be injected is an interface or class coming from a 3rd party library that cannot be modified (in case you wonder, I am having beef with org.apache.http.client.HttpClient).

In other words, I find myself missing (horribly) the feature of accessing Injection Points from @Provides methods that CDI provides ( https://docs.jboss.org/weld/reference/latest/en-US/html/injection.html ). One may also think if it makes sense to provide (eh) a new type of Provider that accepts the InjectionPoint as argument. Alternatively, in order not to introduce another interface with all the likely massive rework it would require in the codebase, if the provider has a field annotated with, say, @Inject@TargetInjectionPoint (where @TargetInjectionPoint would be a binding annotation provided by Guice), that field would be injected with the target Injection Point right before the Providder.get() method is invoked.

**Comment by mmanciop** [comment]
Your library looks good! Not being able to set values in BindingAnnotations is not too much of a limitation, those values can be conveyed by annotations that are not BindingAnnotations. (In any case, there can be only one BindingAnnotation on any InjectionPoint, so that and names exclude one another, which is a downside).

**Comment by tavianator** [comment]
Thanks!  Actually I oversimplified it a bit, the code I gave will only work for the case where the member is a `Field`.  If it's a `Method` or `Constructor`, you'll get the annotations on the method, not the parameter.  The fix is to look for yourself in `injectionPoint.getDependencies()`.  And now that I think about it, maybe my library should just hand you the `Dependency` instead of the `InjectionPoint`.  Perhaps I'll change that in version 2.

**Comment by lsamayoa** [comment]
I guess this will never be part of guice :/
